src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,240 @@
+                beginning, useful to make data start at the beginning of the day or
+                hour.
+        """
+        self.buffer = buffer
+        self.resolution_in_seconds = resolution_in_seconds
+        self.window_start = window_border
+
+        self.missing_windows = []
+        self.datetime_newest = datetime.min
+        self.datetime_oldest = datetime.max
+
+    def max_size(self):
+        """Return the max amount of items this container can hold."""
+        return len(self.buffer)
+
+    def update(self, datetime: datetime, value: T, missing: bool = False):
+        """Update the buffer with a new value for the given timestamp.
+
+        Args:
+            datetime: Timestamp of the new value
+            value: value to add
+            missing: if true, the given timestamp will be recorded as missing.
+                The value will still be written.
+        """
+        # Update timestamps
+        if self.datetime_newest < datetime:
+            self.datetime_newest = datetime
+
+        if self.datetime_oldest > datetime:
+            self.datetime_oldest = datetime
+        elif self.datetime_oldest < self.datetime_newest - timedelta(
+            seconds=len(self.buffer) * self.resolution_in_seconds
+        ):
+            self.datetime_oldest = self.datetime_newest - timedelta(
+                len(self.buffer) * self.resolution_in_seconds
+            )
+
+        # Update data
+        insert_index = self.datetime_to_index(datetime)
+
+        self.buffer[insert_index] = value
+
+        # Update list of missing windows
+        #
+        # We always append to the last pending window.
+        # A window is pending when end == None
+        if missing:
+            # Create new if no pending window
+            if len(self.missing_windows) == 0 or self.missing_windows[-1].end != None:
+                self.missing_windows.append({"start": datetime, "end": None})
+        elif len(self.missing_windows) > 0:
+            # Finalize a pending window
+            if self.missing_windows[-1].end == None:
+                self.missing_windows[-1].end = datetime
+
+        # Delete out-to-date windows
+        if len(self.missing_windows) > 0 and self.missing_windows[0].end != None:
+            if self.missing_windows[0].end <= self.datetime_oldest:
+                self.missing_windows = self.missing_windows[1:]
+
+    def datetime_to_index(self, datetime: datetime) -> int:
+        """Convert the given timestamp to an index.
+
+        Throws an index error when the timestamp is not found within this
+        buffer.
+
+        Args:
+            datetime: Timestamp to convert.
+        Returns:
+            index where the value for the given timestamp can be found.
+        """
+        if datetime < self.datetime_oldest:
+            raise IndexError()
+
+        return self.wrap(int(abs((self.window_start - datetime).total_seconds())))
+
+    def window(self, start: datetime, end: datetime) -> Any:
```

Comment:
Maybe return `Sequence[T]` instead (try to avoid `Any` whenever possible as it kills type checking):
```suggestion
    def window(self, start: datetime, end: datetime) -> Sequence[T]:
```